[BEAM-6986] TypeHints Py3 Error: Multiple tests fail with a generator StopIteration error on Python 3.7

NikeNano opened the pull request:
Update to solve the problem with "Multiple tests fail with a generator StopIteration error on Python 3.7". The solution is based upon using a for loop instead of next where "The for loop listens for StopIteration explicitly." Please see this answer for a good explanation: https://stackoverflow.com/questions/14413969/why-does-next-raise-a-stopiteration-but-for-do-a-normal-return Also of interest is the following:  Change StopIteration handling inside generators https://www.python.org/dev/peps/pep-0479/. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

NikeNano commented:
R: @aaltay, @RobbeSneyders, @tvalentyn 

Always happy to get as much feedback, tips to improve as possible, thanks :)

tvalentyn commented:
@NikeNano Thanks for this fix! 
I suggest we wait until https://github.com/apache/beam/pull/8500 is merged, in the mean time:
1. Let's remove the skip annotation https://github.com/apache/beam/blob/e4cf63b9c69cdd8ce37208bb301595f5f2af12c8/sdks/python/apache_beam/typehints/typehints_test.py#L714
2. Please squash the commits as this is ready for merge.
3. Please add a description to this PR that explains the nature of the change in case someone looks at this PR later and tries to understand the reasoning behind the change. You could add something like: "In Python 3.6 `next()` behaves so and so, however in Python 3.7 it's so and so (see: http://some.link)."

tvalentyn commented:
cc: @Juta, @fredo838 who are also helping with Python 3 support for visibility.

NikeNano commented:
1) I might be misstaken but I believe it is already removed: https://github.com/NikeNano/beam/blob/generator_stopitter/sdks/python/apache_beam/typehints/typehints_test.py.

2) I have squashed the commits. 

3) Will update it :)

tvalentyn commented:
The reason you don't see the `@skip` annotation on your branch is because your branch is older than the commit that added this annotation. You could rebase the change on top of most recent master.

Also note that the PR still has 3 commits, so they are not yet squashed: https://github.com/apache/beam/pull/8497/commits.

You could pull recent master branch, rebase your change on top of master, and combine the commits using squash or fixup. I usually run `git pull --rebase=interactive origin master` to do all those steps in one command, but you can also separate `git pull origin master` and `git rebase -i origin master`.

You may need to do a force-push afterwards.

NikeNano commented:
Finally, now it should be updated (and squashed). Is it standard to squash all commits to 1, before you merge it?

robertwb commented:
Thanks for the change--looks like this was a good cleanup in general. 

It's standard to squash commits unless they have individual significance and value.

NikeNano commented:
@robertwb awesome, thanks!

tvalentyn commented:
LGTM, thanks, @NikeNano .
As for squashing commits, I recommend to squash commits before you submit commits for review, or once the pull request is reviewed and ready for merge, but try to avoid force-pushes once you have created the PR and it is being reviewed. If you are addressing reviewer's feedback I'd recommend to make fixup commits, rather than squashing your change into the first commit. Otherwise it becomes harder to see what has changed between review iterations. Instead, you can add a commit with message that starts with "fixup:", for example, "fixup:  Fix linter."

tvalentyn commented:
@robertwb or @pabloem,  could you please help with the merge? Thank you!

NikeNano commented:
> LGTM, thanks, @NikeNano .
> As for squashing commits, I recommend to squash commits before you submit commits for review, or once the pull request is reviewed and ready for merge, but try to avoid force-pushes once you have created the PR and it is being reviewed. If you are addressing reviewer's feedback I'd recommend to make fixup commits, rather than squashing your change into the first commit. Otherwise it becomes harder to see what has changed between review iterations. Instead, you can add a commit with message that starts with "fixup:", for example, "fixup: Fix linter."

Sounds good, will make it to a habit :)

tvalentyn commented:
As @robertwb said, it is also perfectly fine to have multiple commits if they can be regarded as independent meaningful changes of their own.

pabloem commented:
Thanks @NikeNano et al : )
